---
title: "BrainSeq-preprocessing"
author: "Urwah Nawaz"
date: "2023-08-27"
output: 
  html_document:
    code_folding: hide
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    autodep = TRUE,
    eval = TRUE,
	echo = FALSE,
	warning = FALSE,
	message = FALSE
)
```

# Set-up

```{r include = FALSE}
source("functions.R")
source("../signatures/libraries.R")
source("../signatures/functions.R")
load("../../DeconRNAShiny/sigsBrain.rda")
library(recount3)
library(magrittr)
library(tibble)
library(reshape2)
library(SummarizedExperiment)
library(corrplot)
library(dplyr)
library(ggvenn)
library(pander)
library(gridExtra)
library(variancePartition)
library(DT)
library(EnsDb.Hsapiens.v86)
library(singscore)
library(AnnotationHub)
library(stargazer)
library(ggfortify)
library(glue)
library(cowplot)
library(broom)
library(glmpca)
library(naniar)
```


# Background 

# Preprocessing 

Following analyses and steps will be performed before the data is ready to be exported to Brain Integrative Transcriptome Hub:

- Metadata pre-processing including: 
    - Adding brain region nomenclature
    - Adding developmental stages 
    - Consistent labeling of column names
    - Adding metadata from different resources 
    
- Quality control checks including: 
    - Regions labels (Experimental method)
    - Determining 3' bias that can arise as a result of degradation
    - GC content of the datatset 
    - PCA
  
- Cellular deconvolution of data-set and performing goodness-of-fit
  - Deconvolve using MultiBrain

- Determining drivers of variation


## Metdata pre-processing 

```{r}
annot = read.csv("/home/neuro/Documents/Brain_integrative_transcriptome/brain_transcriptome/annotations/BrainSeq-metadata-annot.csv", header=TRUE)

dir = file.path("/home/neuro/Documents/BrainData/Bulk/Brainseq")

```


```{r}
load(file.path(dir, "rse_gene_unfiltered.Rdata"), envir = .GlobalEnv)
load(file.path(dir,"methprop_pd.Rdata"), envir = .GlobalEnv)
x = rse_gene@colData 
x <- as.data.frame(x)
x <- as.data.frame(t(x))
replicated <- colnames(x)[grep(",", x["SAMPLE_ID",])]
y <- as.list(x)
y[replicated] <- lapply(y[replicated], function(z) {
  # which variables to merge
  to.weight <- which(sapply(z, length) > 1 & sapply(z, class) %in% c("numeric", "integer"))
  # weighting of the merge
  weighting <- z$numReads # total reads
  weighting <- weighting / sum(weighting) # rather than a straight average, it's based on the number of reads
  
  # apply weighting
  z[to.weight] <- lapply(z[to.weight], function(zz) {
    if (length(weighting) == length(zz)) {
      sum(weighting * zz)
      } else {
          NaN
        }
        
      })
      
      # quickly fix character variables
      char <- which(sapply(z, length) > 1 & sapply(z, class) == "character")
      z[char] <- lapply(z[char], function(zz) {
        paste(zz, collapse = " & ")
      })
      
      return(z)
    })
    
    w <- lapply(y, as.data.frame)
    w <- do.call("rbind", w)
    
    comp <- as.data.frame(pd)
    comp <- comp[,57:64]
    m <- match(rownames(comp), rownames(w)) # they are
    md<- cbind(w, comp[m,])
    colnames(md) = annot$BITColumnName[match(colnames(md), annot$OriginalMetadataColumnName)]
    
    
    # Adding features 
md %<>% mutate(Period = ifelse(.$Age > 0, "Postnatal", "Prenatal"), 
                   StructureAcronym = gsub("HIPPO", "HIP", .$StructureAcronym),
                   Diagnosis = gsub("Schizo", "Schizophrenia", .$Diagnosis)) %>%
      mutate(Regions = add_feature(.$StructureAcronym, regions)) %>% 
      mutate(Age_rounded = as.character(sapply(na.omit(.$AgeNumeric), num_to_round))) %>% as.data.frame() %>%
      mutate(AgeInterval = as.character(add_feature(.$Age_rounded, age_intervals))) %>% 
      dplyr::select(-Age_rounded) %>%
      dplyr::select("SampleID", everything()) %>%
      as.data.frame()
```


```{r exp-gene}
exp = rse_gene@assays@.xData$data$rpkm
rownames(exp) <- sub("\\.[0-9]*$", "", rownames(exp))
```


### Export metadata and expression data 

```{r}
write.csv(exp, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-exp.csv")
write.csv(md, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-metadata.csv")
```

## Quality control checks 

- Check sample labeling 


- Determining 3' bias that can arise as a result of degradation

```{r}
#read coverage h (average RPKM Ã— L) reflects the abundance and length of each transcript, but not the sequencing #depth of each sample.


```


- GC content of the dataset 

```{r}


```



# Deconvolution 

```{r}
bseq.decon = run_dtg(exp ,sigsBrain$MB) 

gof_res = write.gof(exp, bseq.decon, 
                    signatureUsed = sigsBrain$MB)


gof_res %>% 
    mutate(col = "BrainSeq") %>%
    as.data.frame() %>% 
    rownames_to_column("SampleID") %>% 
    mutate(SampleID = gsub("X", "", SampleID)) %>%
  mutate(SampleID = gsub("\\.", "_", SampleID)) %>%
    left_join(md, by = "SampleID") %>%
    drop_na(Regions) %>%
    ggplot(aes(Regions, r, fill = Regions)) + geom_violin(width=1, alpha = 0.85) + 
  geom_boxplot(width=0.1, color="white") +
  theme_bw() + scale_fill_manual(values = c("Cortex" = "#E85772", 
                                             "Subcortex" = "#878DC5")) + 
  geom_hline(yintercept = c(0.5,0.7), color = "grey",linetype='dotted') + 
  ylab("Goodness of fit (r)")
```

```{r}
gof_res %>% 
    mutate(col = "BrainSeq") %>%
    as.data.frame() %>% 
    rownames_to_column("SampleID") %>% 
    mutate(SampleID = gsub("X", "", SampleID)) %>%
  mutate(SampleID = gsub("\\.", "_", SampleID)) %>%
    left_join(md, by = "SampleID") %>%
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
    drop_na(Regions) %>%
    ggplot(aes(Period, r, fill = Period)) + geom_violin(width=1) + 
  geom_boxplot(width=0.1, color="white") +
  theme_bw() + scale_fill_manual(values = c("Prenatal" = "#F38BA9", 
                                             "Postnatal" = "#46597D")) + 
  geom_hline(yintercept = c(0.5,0.7), color = "grey",linetype='dotted') + 
  ylab("Goodness of fit (r)")
```


Comparison with other deconvolution 

```{r}
decon.comp = bseq.decon %>% 
  as.data.frame() %>%
  rownames_to_column("SampleID") %>% 
  left_join(md %>% 
              dplyr::select(SampleID, 
                            "Dev.Replicating", 
                            "Dev.Quiescent", 
                            "Adult.OPC", 
                            "Adult.Neurons", 
                            "Adult.Astrocytes", 
                            "Adult.Oligo", 
                            "Adult.Microglia", 
                            "Adult.Astrocytes",
                            "Regions", 
                             "Period",
                            "AgeInterval"))

decon.comp %>% 
  dplyr::select(SampleID, "Dev.Quiescent", "Dev.Replicating","Adult.Neurons", "Neurons", 
                "Regions", "Period", "AgeInterval") %>% melt() %>% 
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
  drop_na() %>%
  ggplot(aes(variable,value, fill = variable)) +
   geom_boxplot(width=0.1) + facet_wrap(~Period) + geom_jitter(alpha=0.5)
```
```{r}
decon.comp %>% 
  dplyr::select(SampleID, "Dev.Quiescent", "Dev.Replicating","Adult.Astrocytes", "Astrocytes", 
                "Regions", "Period", "AgeInterval") %>% melt() %>% 
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
  drop_na() %>%
  ggplot(aes(variable,value, fill = variable)) +
   geom_boxplot(width=0.1) + facet_wrap(~Period) + geom_jitter(alpha=0.5)

```