---
title: "BrainSeq-preprocessing"
author: "Urwah Nawaz"
date: "2023-08-27"
output: 
  html_document:
    code_folding: hide
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    autodep = TRUE,
    eval = TRUE,
	echo = FALSE,
	warning = FALSE,
	message = FALSE
)
```

# Set-up

```{r include = FALSE}
source("functions.R")
source("../signatures/libraries.R")
source("../signatures/functions.R")
load("../../DeconRNAShiny/sigsBrain.rda")
library(recount3)
library(magrittr)
library(tibble)
library(reshape2)
library(SummarizedExperiment)
library(corrplot)
library(dplyr)
library(ggvenn)
library(pander)
library(gridExtra)
library(variancePartition)
library(DT)
library(EnsDb.Hsapiens.v86)
library(singscore)
library(AnnotationHub)
library(stargazer)
library(ggfortify)
library(glue)
library(cowplot)
library(broom)
library(glmpca)
library(naniar)
```


# Background 

# Preprocessing 

Following analyses and steps will be performed before the data is ready to be exported to Brain Integrative Transcriptome Hub:

- Metadata pre-processing including: 
    - Adding brain region nomenclature
    - Adding developmental stages 
    - Consistent labeling of column names
    - Adding metadata from different resources 
    
- Quality control checks including: 
    - Regions labels (Experimental method)
    - Determining 3' bias that can arise as a result of degradation
    - GC content of the datatset 
    - PCA
  
- Cellular deconvolution of data-set and performing goodness-of-fit
  - Deconvolve using MultiBrain

- Determining drivers of variation


## Metdata pre-processing 

```{r}
annot = read.csv("/home/neuro/Documents/Brain_integrative_transcriptome/brain_transcriptome/annotations/BrainSeq-metadata-annot.csv", header=TRUE)

dir = file.path("/home/neuro/Documents/BrainData/Bulk/Brainseq")

```


```{r}
load(file.path(dir, "rse_gene_unfiltered.Rdata"), envir = .GlobalEnv)
load(file.path(dir,"methprop_pd.Rdata"), envir = .GlobalEnv)
x = rse_gene@colData 
x <- as.data.frame(x)
x <- as.data.frame(t(x))
replicated <- colnames(x)[grep(",", x["SAMPLE_ID",])]
y <- as.list(x)
y[replicated] <- lapply(y[replicated], function(z) {
  # which variables to merge
  to.weight <- which(sapply(z, length) > 1 & sapply(z, class) %in% c("numeric", "integer"))
  # weighting of the merge
  weighting <- z$numReads # total reads
  weighting <- weighting / sum(weighting) # rather than a straight average, it's based on the number of reads
  
  # apply weighting
  z[to.weight] <- lapply(z[to.weight], function(zz) {
    if (length(weighting) == length(zz)) {
      sum(weighting * zz)
      } else {
          NaN
        }
        
      })
      
      # quickly fix character variables
      char <- which(sapply(z, length) > 1 & sapply(z, class) == "character")
      z[char] <- lapply(z[char], function(zz) {
        paste(zz, collapse = " & ")
      })
      
      return(z)
    })
    
    w <- lapply(y, as.data.frame)
    w <- do.call("rbind", w)
    
    comp <- as.data.frame(pd)
    comp <- comp[,57:64]
    m <- match(rownames(comp), rownames(w)) # they are
    md<- cbind(w, comp[m,])
    colnames(md) = annot$BITColumnName[match(colnames(md), annot$OriginalMetadataColumnName)]
    
    
    # Adding features 
md %<>% mutate(Period = ifelse(.$Age > 0, "Postnatal", "Prenatal"), 
                   StructureAcronym = gsub("HIPPO", "HIP", .$StructureAcronym),
                   Diagnosis = gsub("Schizo", "Schizophrenia", .$Diagnosis)) %>%
      mutate(Regions = add_feature(.$StructureAcronym, regions)) %>% 
      mutate(Age_rounded = as.character(sapply(na.omit(.$AgeNumeric), num_to_round))) %>% as.data.frame() %>%
      mutate(AgeInterval = as.character(add_feature(.$Age_rounded, age_intervals))) %>% 
      dplyr::select(-Age_rounded) %>%
      dplyr::select("SampleID", everything()) %>%
      as.data.frame()
```


```{r exp-gene}
exp = rse_gene@assays@.xData$data$rpkm
rownames(exp) <- sub("\\.[0-9]*$", "", rownames(exp))
```


### Export metadata and expression data 

```{r}
write.csv(exp, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-exp.csv")
write.csv(md, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-metadata.csv")
```

## Quality control checks 

```{r}
bseq.exp.filt <- thresh(exp)
```

```{r}
pc = log(bseq.exp.filt + 0.05) %>% 
  t() %>% 
   prcomp(scale = TRUE)

pca_bseq =pc$x %>% 
  as.data.frame() %>%
  dplyr::select(PC1, PC2, PC3) %>%
   rownames_to_column("SampleID") %>% 
  left_join(md, by = "SampleID") %>%
 # pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>% 
  ggplot(
    aes(PC1, PC2, colour = mito_Rate, shape = StructureAcronym)
  ) +
  geom_point(size=5) + theme_bw() +
   labs(
    x = glue("PC1 ({percent(pc$sdev[[1]]^2 / sum(pc$sdev^2), 0.1)})"),
    y = glue("PC2 ({percent(pc$sdev[[2]]^2 / sum(pc$sdev^2), 0.1)})"),
    colour = "Mitochondrial rate",
    shape = "Structure Acronym"
  ) + theme(legend.position = "top")


 pc$x  %>%
    as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  left_join(md) %>%
    dplyr::select(
        PC1, PC2, PC3, 
        Period, 
        Regions, 
        AgeNumeric,
        Sex, 
        DonorID,
        RIN, 
        mito_Rate, 
        TotalNReads, 
        MappingRate
    ) %>% 
  drop_na(Regions) %>%
   mutate(Period = as.numeric(as.factor(Period)), 
          Regions = as.numeric(as.factor(Regions)),
          Sex = as.numeric(as.factor(Sex)), 
          DonorID = as.numeric(as.factor(DonorID))) %>%
    cor() %>% 
    corrplot(
        type = "lower", 
        diag = FALSE, 
        addCoef.col = 1, 
        number.cex = 0.75)


ggsave(cor_plot,
       filename = "/home/neuro/Documents/Brain_integrative_transcriptome/Results/Thesis_plots/corplot_bseq.svg", 
        height = 6, 
       width = 8, 
       units = "in")

ggsave(pca_bseq,
       filename = "/home/neuro/Documents/Brain_integrative_transcriptome/Results/Thesis_plots/pca_bseq.svg", 
        height = 6, 
       width = 8, 
       units = "in")
```
- Check sample labeling 


- Determining 3' bias that can arise as a result of degradation

```{r}
#read coverage h (average RPKM Ã— L) reflects the abundance and length of each transcript, but not the sequencing #depth of each sample.


```


- GC content of the dataset 

```{r}
annot_seq = annot %>% 
  dplyr::filter(Type == "Sequencing metrics")
md.clean = md %>% 
  dplyr::select(SampleID, contains(annot_seq$BITColumnName))
md.clean = md.clean %>% select_if(~ !any(is.na(.)))
md.clean = md.clean[vapply(md.clean, function(x) length(unique(x)) > 1, logical(1L))]
```

- Correlation of metadata variables 

```{r}
M = cor(data.matrix(md.clean), use = "complete.obs")
corrplot(M, order='AOE',
         method = "circle", type = "upper", 
         number.cex = .50, 
         tl.cex=0.75)


```


## Determing drivers of variation 



```{r fig.height =5, fig.width=5, fig.cap="*Assessing correlation between covariates of interest from the BrainSeq data*"}
form.bseq <- ~ AgeNumeric + (1|StructureAcronym) + (1|Sex) + RIN +  (1|Diagnosis) + mito_Rate + rRNA_rate + TotalNReads + MappingRate + 
  (1|Period) + (1|AgeInterval) + (1|DonorID)
C.bseq = canCorPairs(form.bseq, md)
plotCorrMatrix( C.bseq )

```

The correlation plot reveals that many of the selected covariates do not correlate highly within their respective category, and therefore we will feed these into the mixed linear model. 


```{r}
varPar.bseq <- fitExtractVarPartModel(bseq.exp.filt, form.bseq, md)
```


```{r fig.width=10, fig.cap="*variancePartition results for BrainSeq. Plot on right shows impact of covariates of interest on randomly selected genes whereaas left shows overall impact of covariates on all gene expression.*"}

var.genes.bseq = plotVarPart(varPar.bseq %>% 
                               as.data.frame() %>% 
                               dplyr::select( Period, mito_Rate,
                                             StructureAcronym,
                                             DonorID, AgeInterval, 
                                             RIN, AgeNumeric,MappingRate, 
                                             rRNA_rate,
                                             everything()))
var.genes.bseq <- variancePartition::sortCols(var.genes.bseq)

ggsave(var.genes.bseq, 
       filename = "/home/neuro/Documents/Brain_integrative_transcriptome/Results/Thesis_plots/varPart_bseq.svg",
       height = 6, 
       width = 8, 
       units = "in")

grid.arrange(var.genes.bseq, var.total.bseq , ncol = 2)

write.csv(varPar.bseq, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-varPart.csv")
```


# Deconvolution 

```{r}
bseq.decon = run_dtg(exp ,sigsBrain$MB) 

gof_res = write.gof(exp, bseq.decon, 
                    signatureUsed = sigsBrain$MB)


gof_res %>% 
    mutate(col = "BrainSeq") %>%
    as.data.frame() %>% 
    rownames_to_column("SampleID") %>% 
    mutate(SampleID = gsub("X", "", SampleID)) %>%
  mutate(SampleID = gsub("\\.", "_", SampleID)) %>%
    left_join(md, by = "SampleID") %>%
    drop_na(Regions) %>%
    ggplot(aes(Regions, r, fill = Regions)) + geom_violin(width=1, alpha = 0.85) + 
  geom_boxplot(width=0.1, color="white") +
  theme_bw() + scale_fill_manual(values = c("Cortex" = "#E85772", 
                                             "Subcortex" = "#878DC5")) + 
  geom_hline(yintercept = c(0.5,0.7), color = "grey",linetype='dotted') + 
  ylab("Goodness of fit (r)")
```

```{r}
gof_res %>% 
    mutate(col = "BrainSeq") %>%
    as.data.frame() %>% 
    rownames_to_column("SampleID") %>% 
    mutate(SampleID = gsub("X", "", SampleID)) %>%
  mutate(SampleID = gsub("\\.", "_", SampleID)) %>%
    left_join(md, by = "SampleID") %>%
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
    drop_na(Regions) %>%
    ggplot(aes(Period, r, fill = Period)) + geom_violin(width=1) + 
  geom_boxplot(width=0.1, color="white") +
  theme_bw() + scale_fill_manual(values = c("Prenatal" = "#F38BA9", 
                                             "Postnatal" = "#46597D")) + 
  geom_hline(yintercept = c(0.5,0.7), color = "grey",linetype='dotted') + 
  ylab("Goodness of fit (r)")
```


Comparison with other deconvolution 

```{r}
decon.comp = bseq.decon %>% 
  as.data.frame() %>%
  rownames_to_column("SampleID") %>% 
  left_join(md %>% 
              dplyr::select(SampleID, 
                            "Dev.Replicating", 
                            "Dev.Quiescent", 
                            "Adult.OPC", 
                            "Adult.Neurons", 
                            "Adult.Astrocytes", 
                            "Adult.Oligo", 
                            "Adult.Microglia", 
                            "Adult.Astrocytes",
                            "Regions", 
                             "Period",
                            "AgeInterval"))

decon.comp %>% 
  dplyr::select(SampleID, "Dev.Quiescent", "Dev.Replicating","Adult.Neurons", "Neurons", 
                "Regions", "Period", "AgeInterval") %>% melt() %>% 
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
  drop_na() %>%
  ggplot(aes(variable,value, fill = variable)) +
   geom_boxplot(width=0.1) + facet_wrap(~Period) + geom_jitter(alpha=0.5)
```


```{r}
decon.comp %>% 
  dplyr::select(SampleID, "Dev.Quiescent", "Dev.Replicating","Adult.Astrocytes", "Astrocytes", 
                "Regions", "Period", "AgeInterval") %>% melt() %>% 
  mutate(Period = factor(Period, levels = c("Prenatal", "Postnatal"))) %>%
  drop_na() %>%
  ggplot(aes(variable,value, fill = variable)) +
   geom_boxplot(width=0.1) + facet_wrap(~Period) + geom_jitter(alpha=0.5)

```

### Export 

```{r}
write.csv(bseq.decon, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-decon.csv")

```