---
title: "Recount exploration"
author: "Urwah Nawaz"
date: "2023-06-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    autodep = TRUE,
    eval = TRUE,
	echo = FALSE,
	warning = FALSE,
	message = FALSE
)
```

# Setup 

```{r}
source("functions.R")
library(dplyr)
library(purrr)
library(recount3)
library(EnsDb.Hsapiens.v86)
library(ggpubr)
library(DT)
library(pheatmap)
library(org.Hs.eg.db)
```

```{r}
ft_cat_annot = read.table("/home/neuro/Documents/Brain_integrative_transcriptome/brain_transcriptome/annotations/gene.info.tsv", sep = "\t", header=TRUE, 
                          fill=TRUE) %>% 
  as.data.frame() %>% 
  dplyr::select(gene.name, gene.category) %>% 
  dplyr::rename(gene_id = gene.name, gene_biotype = gene.category) %>% 
  mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "")
txdf = genes(EnsDb.Hsapiens.v86, return.type="DataFrame")
tx2gene = as.data.frame(txdf[,c("gene_id", "gene_biotype")]) %>% 
  rbind(ft_cat_annot)

```

```{r}
gtex.bH.exp = read.csv("/home/neuro/Documents/BrainData/Bulk/GTEx/Formatted/GTEx-exp.csv", header=TRUE, row.names = 1)
gtex.bH.md.subset = read.csv("/home/neuro/Documents/BrainData/Bulk/GTEx/Formatted/GTEx-metadata-subset.csv", header = TRUE)
```

# Exploring the GTEx dataset from Recount3

This documentation contains the summary of datasets that are currently hosted on BITHub. It also includes levels of information (gene, transcript, exon) and overall summary statistics. 

```{r}
gtex = recount3::create_rse_manual(
    project = "BRAIN",
    project_home = "data_sources/gtex",
    organism = "human",
    annotation = "gencode_v26",
    type = "gene"
)

gtex_fantom = recount3::create_rse_manual(
    project = "BRAIN",
    project_home = "data_sources/gtex",
    organism = "human",
    annotation = "fantom6_cat",
    type = "gene"
)

```


## Metadata variables from Recount 

```{r}
md_rc = colData(gtex)

colnames(md_rc) %>% 
  melt() %>%
   DT::datatable(caption= "Metadata variables from Recount with gencode annotations")

```


```{r}
md_rc = colData(gtex_fantom)

colnames(md_rc) %>% 
  melt() %>%
  DT::datatable(caption= "Metadata variables from Recount with Fantom-CAT annotations")

```


## Annotations from Recount using FANTOM-CAT

```{r}
gtex_recount= assay(gtex, "raw_counts")
dim(gtex_recount)

```



```{r fig.height=10, fig.width=10, fig.cap="Annotations of GTEx genes retrieved via recount3 using Gencode"} 
frac = gtex_recount %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = frac)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Frequency") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

count = gtex_recount %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = n)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Count") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

ggarrange(count,frac, ncol = 2)
```

```{r fig.height=10, fig.width=10}
gtex_recount_fc= assay(gtex_fantom, "raw_counts")
dim(gtex_recount_fc)

```

```{r fig.height=10, fig.width=10,fig.cap="Annotations of GTEx genes retrieved via recount3 using Gencode"}
frac = gtex_recount_fc %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = frac)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Frequency") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

count = gtex_recount_fc %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = n)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Count") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

ggarrange(count,frac, ncol = 2)
```


### Genes similar between Gencode and FANTOM-CAT 

```{r}
library(VennDiagram)
library(ggvenn)
gtex_recount_fc %<>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
  column_to_rownames("gene_id")

gtex_recount %<>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>%
  column_to_rownames("gene_id")

ggvenn(list("Gencode" =rownames(gtex_recount), 
                    "FANTOM-CAT" = rownames(gtex_recount_fc)))
```

* What genes are missing from each data-frame? 

```{r}
overlap_annotations = calculate.overlap(x=list("Gencode" =rownames(gtex_recount), 
                    "FANTON-CAT" = rownames(gtex_recount_fc)))

gencode_only = overlap_annotations$a1[overlap_annotations$a1 != overlap_annotations$a3] %>% 
  as.data.frame() %>%
  set_colnames("gene_id") %>% 
  mutate(category ="Gencode Only") %>%
  mutate(symbol = mapIds(org.Hs.eg.db, keys=.$gene_id,
                             column=c("SYMBOL"),keytype="ENSEMBL",
                         multiVals="first"), 
         ontology = mapIds(org.Hs.eg.db, keys=.$gene_id,
                             column=c("ONTOLOGY"),keytype="ENSEMBL",
                         multiVals="first")) %>% 
  left_join(tx2gene, by ="gene_id")

gencode_only %>% 
  DT::datatable(caption="Ensembl ID genes only found in Gencode annotations")
  
```
```{r fig.height=10, fig.cap="Gene types in Gencode only annotations"}
gencode_only %>% 
   dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = n)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Count") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

```


```{r}
ft_cat_only = overlap_annotations$a2[overlap_annotations$a2 != overlap_annotations$a3] %>% 
  as.data.frame() %>%
  set_colnames("gene_id") %>% 
  mutate(category ="FT-CAT Only") %>%
  mutate(symbol = mapIds(org.Hs.eg.db, keys=.$gene_id,
                             column=c("SYMBOL"),keytype="ENSEMBL",
                         multiVals="first"), 
         ontology = mapIds(org.Hs.eg.db, keys=.$gene_id,
                             column=c("ONTOLOGY"),keytype="ENSEMBL",
                         multiVals="first")) %>% 
  left_join(tx2gene, by ="gene_id")

ft_cat_only  %>% 
  DT::datatable(caption="Ensembl ID genes only found in FANTOM-CAT annotations")

```

```{r fig.height=10, fig.cap="Gene types in FANTOM-CAT Only annotations"}

ft_cat_only  %>% 
   dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = n)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Count") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))
```

## Data for BITHub for GTEx

```{r fig.height=10, fig.width=10, fig.cap="Annotations of GTEx genes retrieved via GTEx portal preprocessed for BITHub"}
frac = gtex.bH.exp %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = frac)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Frequency") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

count = gtex.bH.exp %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>% 
    mutate_at(.vars = "gene_id", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") %>% 
    left_join(tx2gene, by ="gene_id") %>% 
  dplyr::count(gene_biotype) %>%
    mutate(sum=sum(n),frac = n / sum(n)) %>%
  ggplot(aes(x = as.factor(gene_biotype), y = n)) +
    geom_bar( stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -90)) +
    labs(x = "Gene biotype (Ensembl v86)", y = "Count") +
theme(panel.spacing = unit(1, "lines")) +
theme(strip.background = element_rect(fill = "black"))+
theme(strip.text = element_text(color = "white", face = "bold"))

ggarrange(count,frac, ncol = 2)
```


# Additional quality control metrics using Recount3

* Megadepth: Computes coverage across a bigwig file (which has been provided by Recount3)

```{r}

```
<!-- ```{r} -->
<!-- colnames(exp) <- gsub("\\.", "-", colnames(exp)) # Changing expression file names to match metadata SampleIDs -->
<!-- exp %<>% column_to_rownames("Name") -->
<!-- exp %>%  -->
<!--       dplyr::select(contains(md$SampleID)) -->
<!--     message(paste0("Samples subsetted - Exp matrix contains ", ncol(exp), " samples")) -->
<!--     rownames(exp) <- sub("\\.[0-9]*$", "", rownames(exp)) -->

<!--     md= md[which(md$SampleID %in% colnames(exp)),]  -->

<!-- ``` -->

<!-- # ```{r} -->
<!-- # org = recount3::create_rse_manual( -->
<!-- #     project = "ERP109002", -->
<!-- #     project_home = "data_sources/sra", -->
<!-- #     organism = "human", -->
<!-- #     annotation = "gencode_v26", -->
<!-- #     type = "gene" -->
<!-- # ) -->
<!-- # ``` -->
<!-- #  -->
<!-- #  -->
<!-- # ```{r} -->
<!-- # org_md = colData(org) -->
<!-- # org_md %>% as.data.frame() -->
<!-- # org_md$sra.library_name -->
<!-- # ``` -->