---
title: "Metadata correlations and determining drivers of variation"
author: "Urwah Nawaz"
date: "2023-03-22"
output: 
  html_document:
    code_folding: hide

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Background 

BITHub has two aims 

1) Allowing the comparison of expression of gene (or genes) of interest across multiple datasets 

2) Allowing the comparison of expression against metadata variables in a given data-set

For the second aim, it is vital that BITHub contains the relevant metadata annotations.
BITHub aims to provide 3 types of metadata annotations for each dataset in the web-browser: 

* Phenotype annotations: 


* Sequencing metrics 

* Sample characteristics 
These annotations relate to how the samples were experimentally prepared

In order to ensure the metadata information is displayed in a user-friendly, highly correlating metadata annotations will be removed and a subset will be used for the site. Additionally, we will also perform varianceParition analysis on the subsetted list. 



# Set-up 

```{r}
source("functions.R")
library(pander)
```


# Metedata correlations - Bulk data

## Datasets

### BrainSeq 


```{r fig.height=8, fig.width=12, fig.cap="*Correlation plot of metadata annotations from BrainSeq phase II. The metadata annotations are clustered based on correlation*"}
bseq = read.csv("/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-metadata.csv", header = TRUE, row.names = 1)
bseq %<>% 
  dplyr::select(-c(FQCbasicStats, perSeqQual, SeqLengthDist,KmerContent))
M = cor(data.matrix(bseq), use = "complete.obs")
corrplot(M, order = 'AOE')


```

Prior to running `cor()` function, the FQCbasicStats, perSeqQual, SeqLengthDist and KmerContent columns were removed as they contained the same value, resulting in NA. 

BrainSeq metadata annotations shows duplicate information in many of its columns (e.g SampleID, SAMPLEID), which are likely a result of running the pre-processing pipeline for BITHub. Additionally, certain columns contain very similar information thus resulting in high correlation. Several QC metrics for RNA-seq QC also provide redundant information and they will be removed for downstream analysis. 

The final BrainSeq annotations will contain the following columns: 

```{r}
bseq.annot = read.csv("../annotations/BrainSeq-metadata-annot.csv") %>%
  dplyr::filter(Include..Yes.No....Interest == "Yes") %>% 
  dplyr::select(-c(Include..Yes.No....Interest))


bseq.annot %>% 
  pander(justify = "lll",
    style = "rmarkdown",
    caption = "BrainSeq metadata annotations that will be used for BITHub ")

bseq %<>% dplyr::select(contains(bseq.annot$BITColumnName))

write.csv(bseq, file = "/home/neuro/Documents/BrainData/Bulk/Brainseq/Formatted/BrainSeq-metadata-subset.csv")
```

### BrainSpan

```{r}
bspan = read.csv("/home/neuro/Documents/BrainData/Bulk/BrainSpan/Formatted/BrainSpan-metadata.csv", header = TRUE, row.names = 1)

M = bspan %>% dplyr::select(-c("Diagnosis")) %>% data.matrix() %>% cor(.,use = "complete.obs")

corrplot(M, order='AOE')
```

BrainSpan metadata annotations contain several duplicate and redundant columns that essentially contain the same information (e.g column_num, Age.x, Braincode). BrainSpan annotations were retrieved from multiple sources and as such, these duplicates are likely a result of different IDs they were stored under. 

The following BrainSpan metadata annotations will be used for BITHub: 

```{r}
bspan.annot = read.csv("../annotations/BrainSpan-metadata-annot.csv") %>%
  dplyr::filter(Include..Yes.No....Interest == "Yes") %>% 
  dplyr::select(-c(Include..Yes.No....Interest))

bspan.annot  %>% 
  pander(justify = "lll",
    style = "rmarkdown",
    caption = "BrainSpan metadata annotations that will be used for BITHub ")

bspan %<>% dplyr::select(contains(bspan.annot$BITColumnName))

write.csv(bspan, file = "/home/neuro/Documents/BrainData/Bulk/BrainSpan/Formatted/BrainSpan-metadata-subset.csv")
```



### GTEx

```{r fig.height=12, fig.width=15}
gtex = read.csv("/home/neuro/Documents/BrainData/Bulk/GTEx/Formatted/GTEx-metadata.csv", header = TRUE, row.names = 1)

M = cor(data.matrix(gtex))

corrplot(M, method = 'number')

```


The GTEx metadata contains comprehensive annotations of sample, sequencing and phenotype attributes. However, for BITHub, we will remove metadata annotations that have a strong positive correlation. 

The following metadata annotations will be used for GTEx: 


```{r}

gtex.annot = read.csv("../annotations/GTEx-metadata-annot.csv") %>%
  dplyr::filter(Include..Yes.No....Interest == "Yes") %>% 
  dplyr::select(-c(Include..Yes.No....Interest))

gtex %<>% dplyr::select(contains(gtex.annot$BITColumnName))

gtex.annot %>% 
  pander(justify = "lll",
    style = "rmarkdown",
    caption = "GTEx metadata annotations that will be used for BITHub ")

write.csv(bspan, file = "/home/neuro/Documents/BrainData/Bulk/GTEx/Formatted/GTEx-metadata-subset.csv")


```


### PsychEncode 

```{r fig.height=8, fig.width=12}
pe = read.csv("/home/neuro/Documents/BrainData/Bulk/PsychEncode/Formatted/PsychEncode-metadata.csv", header = TRUE, row.names = 1)

M = pe %>% 
  dplyr::select(-c(ageBiopsy, smellTestScore,smoker,Structure, StructureAcronym, Regions, Capstone_4, Adult.In7)) %>%
  data.matrix() %>% cor(.,use ='pairwise.complete.obs' )


corrplot(M)
```
Information regarding Row_IDs, Row_Versions, Contributing Studes and Notes will be removed for BITHub. 

The following metadata annotations will be retained for PsychEncode:


```{r}
pe.annot = read.csv("../annotations/PsychEncode-metadata-annot.csv") %>%
  dplyr::filter(Include..Yes.No....Interest == "Yes") %>% 
  dplyr::select(-c(Include..Yes.No....Interest))

pe %<>% dplyr::select(contains(pe.annot$BITColumnName))

pe.annot %>% 
  pander(justify = "llr",
    style = "rmarkdown",
    caption = "PsychEncode metadata annotations that will be used for BITHub ")

write.csv(bspan, file = "/home/neuro/Documents/BrainData/Bulk/GTEx/Formatted/GTEx-metadata-subset.csv")
```

# Determing drivers of variation 

* What is variance partition

## BrainSeq 

* Filtering - how many genes are left and why is this necessary 
* Choosing of metadata variables 

```{r}
#bseq <- read.csv("datasets/FormattedData/FormattedData/BrainSeq/BrainSeq-exp.csv", row.names =2)[,-1]
#ead(bseq)[1:10]
#bseq.exp <- bseq[apply(bseq >= 1, 1, sum) >= 0.1*ncol(bseq),]
```


```{r}
#bseq.md <- read.csv("datasets/FormattedData/FormattedData/BrainSeq/BrainSeq-metadata-subset.csv", row.names = 1)
#head(bseq.md)
```

```{r}
#form.bseq <- ~ AgeNumeric + (1|StructureAcronym) + (1|Sex) + RIN +  (1|Diagnosis) + mito_Rate + rRNA_rate + TotalNReads + MappingRate
#+ Adult.Oligo + Adult.Microglia + Adult.Endothelial
```


```{r}
#varPar.bseq <- fitExtractVarPartModel(bseq.exp, form.bseq, bseq.md)
```


## BrainSpan 

* Filtering 
* Choosing of metadata variables 

```{r}
#bspan.exp = read.csv("/home/neuro/Documents/BrainData/Bulk/BrainSpan/Formatted/BrainSpan-exp.csv", row.names = 1, check.names = FALSE) #%>%
#    column_to_rownames("EnsemblID")
#bspan.meta = read.csv("/home/neuro/Documents/BrainData/Bulk/BrainSpan/Formatted/BrainSpan-metadata.csv")


#bspan.exp <- bspan.exp[apply(bspan.exp >= 1, 1, sum) >= 0.1*ncol(bspan.exp),]
```

```{r}

#form.bspan <- ~ AgeNumeric + (1|StructureAcronym) + (1|Sex) + (1|Period) + (1|Regions)

#form.bspan <- ~ AgeNumeric + (1|StructureAcronym) + (1|Sex) + (1|Period) + (1|Regions) 
```

```{r}
#varPar.bspan <- fitExtractVarPartModel(bspan.exp, form.bspan, bspan.meta)
```


## GTEx 

```{r}
#gtex.exp <- read.csv("datasets/FormattedData/FormattedData/Gtex/GTEx-exp.csv", row.names = 1)
#gtex.md <- read.csv("datasets/FormattedData/FormattedData/Gtex/GTEx-metadata-subset.csv")
#colnames(gtex.md)
#gtex.exp <- gtex.exp[apply(gtex.exp >= 1, 1, sum) >= 0.1*ncol(gtex.exp),]

#gtex.form <- ~ TotalNReads + rRNA_rate + (1|TypeofBatch) + (1|DateofBatch) + (1|BSS_Collection_side_code) + (1|AgeInterval) + (1|Sex) + #(1|Regions) + IntergenicRate + RIN
```

```{r}
#varPar.gtex <- fitExtractVarPartModel(gtex.exp, gtex.form, gtex.md)
#vp <- sortCols(varPar.gtex)
#plotVarPart(vp)
#write.csv(vp, "GTEx-varPart.csv")
```
