---
title: "Recount Metadata pre-processing and correlations"
author: "Urwah Nawaz"
date: "2023-07-10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    autodep = TRUE,
    eval = TRUE,
	echo = FALSE,
	warning = FALSE,
	message = FALSE
)
```

# Setup 

```{r libraries}
library(recount3)
library(DT)
library(dplyr)
library(tidyr)
source("functions.R")
source("../signatures/libraries.R")
source("../signatures/functions.R")
load("../../DeconRNAShiny/sigsBrain.rda")
library(pander)
library(gridExtra)
library(variancePartition)
library(corrplot)
library(edgeR)
library(purrr)
```

```{r}
annot =  read.csv("../annotations/Recount-annotations.csv", header=TRUE, check.names = FALSE, sep = ",")
```

```{r}
strc_acr <- map_df(structure_acronym, ~as.data.frame(.x), .id="id") %>% 
  set_colnames("StructureAcronym")
colnames(strc_acr)[2] = c("Structure")

rgns = map_df(regions, ~as.data.frame(.x), .id="id") %>% 
  set_colnames("Regions")
colnames(rgns)[2] = c("StructureAcronym")


rgns_ftl = map_df(regions_fetal, ~as.data.frame(.x), .id="id") %>% 
  set_colnames("Regions")
colnames(rgns)[2] = c("StructureAcronym")
```

# Background 


Recount3 contains over 70,000 uniformly processed human RNA-seq samples. Recount provides gene, exon and exon-exon junction count matrices both in text format and as a `RangedSummarizedExperiment`. 

The reads from recount were algined with the splice-aware Rail-RNA aligner. To compute the gene count matrices, the mapped reads were quantified with Gencode v25 with hg38 coordinates. 

Unlike traditional quantification methods, recount3 provides base-pair coverage counts. Essentially, these are created in the following manner: 

- Ar each exonic base-pair, the number of reads overlapping at that given base pair is computed. However, as library size is provided in recount3, the coverage counts can be scaled to read counts for a given library size. 

# Datasets 

<!-- ## Ramakar et al (2017) -->

<!-- ```{r} -->
<!-- ramakar = recount3::create_rse_manual( -->
<!--   project = "SRP073813", -->
<!--   project_home = "data_sources/sra", -->
<!--   organism = "human", -->
<!--   annotation = "gencode_v29", -->
<!--   type = "gene" -->
<!-- ) -->


<!-- #ramakar.manual = read.table("/home/neuro/Documents/BrainData/Bulk/Ramakar/sra.gene_sums.SRP073813.G026", header=TRUE) -->

<!-- ramakar.refine = read.table("/home/neuro/Documents/BrainData/Bulk/Ramakar/SRP073813/SRP073813.tsv", sep = "\t", header=TRUE) -->
<!-- ``` -->

<!-- ### Metdata preprocessing  -->
<!-- ```{r sra-dataset} -->
<!-- sra_data = colData(ramakar) %>%  -->
<!--   as.data.frame() %>%  -->
<!--   dplyr::select(contains("sra"))  -->


<!-- md.sra = sra_data %>% -->
<!--   dplyr::select("sra.sample_attributes") %>% -->
<!--   rownames_to_column("SampleID") %>% -->
<!--   separate(sra.sample_attributes, c("Age", "pH", "StructureAcronym", "Diagnosis", "Ethnicity", "Sex", -->
<!--                                     "PMI", "SourceName", "Tissue"), sep="\\|") %>%  -->
<!--   mutate(Age = gsub(".*;;","", Age),  -->
<!--          pH = gsub(".*;;","", pH),  -->
<!--          StructureAcronym = gsub(".*;;","", StructureAcronym), -->
<!--          Diagnosis = gsub(".*;;","", Diagnosis),  -->
<!--          Ethnicity = gsub(".*;;","", Ethnicity),  -->
<!--          Sex = gsub(".*;;","", Sex),  -->
<!--          PMI = gsub(".*;;","", PMI)) %>%  -->
<!--   dplyr::select(-c("SourceName", "Tissue")) %>%  -->
<!--   mutate(Sex = ifelse(Sex == "male", "M", "F"),  -->
<!--          StructureAcronym = gsub("nAcc","NAC", StructureAcronym))  %>%  -->
<!--   mutate(StructureAcronym = gsub("AnCg", "ACC", StructureAcronym)) %>% -->
<!--   mutate(Age_rounded = as.character(sapply(na.omit(as.numeric(.$Age)), num_to_round)),  -->
<!--         Period = c("Postnatal"),  -->
<!--         Age = as.numeric(Age),  -->
<!--         pH = as.numeric(pH),  -->
<!--         PMI = as.numeric(PMI)) %>% -->
<!--   mutate(AgeInterval = add_feature(.$Age_rounded , age_intervals)) -->

<!-- md.sra %<>% -->
<!--   left_join(rgns, by = "StructureAcronym")  -->
<!-- ``` -->


<!-- ```{r} -->
<!-- md.refine.bio = read.table("/home/neuro/Documents/BrainData/Bulk/Ramakar/SRP073813/metadata_SRP073813.tsv", header=TRUE, -->
<!--                            sep = "\t") -->

<!-- md.refine.bio %<>%  -->
<!--   as.data.frame() %>% -->
<!--   mutate(DonorID = gsub("_.*","", refinebio_title))  %>%  -->
<!--   dplyr::rename("SampleID" = "refinebio_accession_code") -->



<!-- ``` -->

<!-- ```{r recount} -->
<!-- md = colData(ramakar) %>%  -->
<!--   as.data.frame() %>% -->
<!--   dplyr::select(-contains("2")) -->
<!-- colnames(md) = annot$BITColumnName[match(colnames(md), annot$RecountName)] -->
<!-- md = full_join(md, md.sra) -->

<!-- md.refine.bio %>% -->
<!--   dplyr::select(SampleID,MetaSRA_age.at.death, DonorID) -->
<!-- md = full_join(md, md.refine.bio) -->
<!-- #md_names = colnames(md) %>% as.data.frame() -->
<!-- #write.csv(md_names, file = "../annotations/Ramaker-2017-metadata-annot.csv") -->
<!-- ``` -->


<!-- ```{r} -->
<!-- md.clean = md %>% select_if(~ !any(is.na(.))) -->
<!-- md.clean = md.clean[vapply(md.clean, function(x) length(unique(x)) > 1, logical(1L))] -->
<!-- ``` -->

<!-- * Selection of metadata variables for subset  -->

<!-- <!-- # ```{r fig.height=30, fig.width=30} --> -->
<!-- <!-- # M = cor(data.matrix(md.clean), use = "complete.obs") --> -->
<!-- <!-- # corrplot(M, order='AOE') --> -->
<!-- <!-- # ``` --> -->
<!-- <!-- #  --> -->
<!-- <!-- # ```{r} --> -->
<!-- <!-- # rowMeans(M, na.rm = TRUE) %>% scale() %>% as.data.frame()  --> -->
<!-- <!-- # ``` --> -->

<!-- <!-- * Selection for variance Partition  --> -->

<!-- <!-- ```{r} --> -->
<!-- <!-- ramakar_annot = read.csv("../annotations/Ramaker-2017-metadata-annot.csv", header=TRUE) %>%  --> -->
<!-- <!--   dplyr::filter(Include..Yes.No....Interest == "Yes") %>%  --> -->
<!-- <!--   dplyr::select(-c(Include..Yes.No....Interest)) --> -->


<!-- <!-- knitr::kable(ramakar_annot) --> -->

<!-- <!-- md %<>% dplyr::select(contains(ramakar_annot$BITColumnName)) --> -->
<!-- <!-- ``` --> -->

<!-- <!-- ```{r fig.height=20, fig.width=20} --> -->
<!-- <!-- M = cor(data.matrix(md), use = "complete.obs") --> -->
<!-- <!-- corrplot(M) --> -->
<!-- <!-- ``` --> -->


<!-- ```{r} -->
<!-- assay(ramakar, "counts") = transform_counts(ramakar, round = TRUE) -->

<!-- ram.tpm = recount::getTPM(ramakar) -->
<!-- ram.tpm %<>% as.data.frame() %>% -->
<!--   dplyr::select(-c(SRR3438888)) -->


<!-- #,SRR3438686, SRR3438689, -->
<!--  #                  SRR3438708, SRR3438710, SRR3438736,  -->
<!--   #                 SRR3438712, SRR3438714, SRR3438716, SRR3438734)) -->

<!-- filter <- rowSums(ram.tpm, na.rm = TRUE) > 0 -->

<!-- ram.tpm.filt = ram.tpm[filter,] %>% as.data.frame() -->
<!-- ram.tpm.filt -->

<!-- ram.tpm.filt <- thresh(ram.tpm.filt) -->
<!-- md %<>%  -->
<!--   dplyr::filter(SampleID %in% colnames(ram.tpm.filt)) -->
<!-- ``` -->

<!-- ### Quality control analysis  -->

<!-- ### Deconvolution  -->

<!-- ### Determining drivers of variation -->

<!-- ```{r} -->
<!-- ram.before = log(ram.tpm + 0.01) %>% -->
<!--     melt() %>% -->
<!--     dplyr::filter(is.finite(value)) %>%  -->
<!--     ggplot(aes(x=value, color = as.character(variable))) + -->
<!--     geom_density() + -->
<!--     ggtitle("Before filtering") + -->
<!--     labs(x = "logTPM", y = "Proportion of Genes") +  -->
<!--     theme_bw() + -->
<!--     theme(legend.position = "none") -->



<!-- ram.after =log(as.matrix(ram.tpm.filt) + 0.01) %>% -->
<!--     melt() %>% -->
<!--     dplyr::filter(is.finite(value)) %>%  -->
<!--     ggplot(aes(x=value, color = as.character(Var2))) + -->
<!--     geom_density() + -->
<!--     ggtitle("After filtering") + -->
<!--     labs(x = "logTPM", y = "Proportion of Genes") +  -->
<!--     theme_bw() + -->
<!--     theme(legend.position = "none") -->

<!-- grid.arrange(ram.before, ram.after, ncol =2) -->
<!-- ``` -->

<!-- - PCA -->

<!-- ```{r} -->
<!-- ## Refine bio  -->
<!-- ramakar.refine %<>% column_to_rownames("Gene") -->
<!-- ram.filt <- thresh(ramakar.refine)  -->
<!-- ``` -->


<!-- ```{r} -->
<!-- pc = log(ram.tpm.filt  + 0.05) %>%  -->
<!--   t() %>%  -->
<!--    prcomp(scale = TRUE) -->

<!-- pc$x %>%  -->
<!--   as.data.frame() %>% -->
<!--   dplyr::select(PC1, PC2, PC3) %>% -->
<!--    rownames_to_column("SampleID") %>%  -->
<!--   left_join(md, by = "SampleID") %>% -->
<!--  # pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>%  -->
<!--   ggplot( -->
<!--     aes(PC1, PC2, colour = Size)  -->
<!--   ) + -->
<!--   geom_point() -->

<!-- ``` -->
<!-- ```{r} -->
<!-- log_gene_cor= cor(log(ram.tpm.filt  + 0.05 )) -->

<!-- log_gene_cor[log_gene_cor==1] <- NA -->

<!-- cor = rowMeans(log_gene_cor, na.rm = TRUE) %>% scale() %>% as.data.frame() %>% -->
<!--   set_colnames("Genes") -->

<!-- cor %>% DT::datatable() -->
<!-- ``` -->


<!-- - Variance partition  -->

<!-- ```{r} -->
<!-- form.ram <- ~ Age + (1|Sex) + (1|Diagnosis) + PMI + pH + TotalNReads + MappingRate + (1|Regions) + (1|DonorID) -->
<!-- C.ram = canCorPairs(form.ram, md) -->
<!-- plotCorrMatrix(C.ram) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- varPar.ram <- fitExtractVarPartModel(ram.tpm.filt, form.ram ,md ) -->

<!-- ``` -->

<!-- ```{r} -->
<!-- var.genes.ram = plotVarPart(varPar.ram %>%  -->
<!--                               as.data.frame() %>%  -->
<!--                               dplyr::select(DonorID, MappingRate,  -->
<!--                                             Regions, TotalNReads, -->
<!--                                             PMI, pH, everything())) -->
<!-- var.total.ram =plotPercentBars(varPar.ram [1:10,] ) -->
<!-- grid.arrange(var.genes.ram, var.total.ram , ncol = 2) -->

<!-- ggsave(var.genes.ram,  -->
<!--        filename = "/home/neuro/Documents/Brain_integrative_transcriptome/Results/Thesis_plots/varPart_ram.svg", -->
<!--        height = 6,  -->
<!--        width = 8,  -->
<!--        units = "in") -->
<!-- ``` -->

<!-- ### Export data  -->

<!-- ```{r} -->
<!-- ram.final = ram.tpm %>%  -->
<!--   as.data.frame() %>%  -->
<!--   rownames_to_column("EnsemblID") %>%  -->
<!--   mutate_at(.vars = "EnsemblID", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "")  -->

<!-- write.csv(ram.final, file= "/home/neuro/Documents/BrainData/Bulk/Ramakar/Formatted/Ramaker-exp.csv") -->

<!-- write.csv(md, file = "/home/neuro/Documents/BrainData/Bulk/Ramakar/Formatted/Ramaker-metadata.csv") -->

<!-- write.csv(varPar.ram, file = "/home/neuro/Documents/BrainData/Bulk/Ramakar/Formatted/Ramaker-varPart.csv") -->

<!-- ``` -->

<!-- ### Deconvolution  -->

<!-- ```{r} -->

<!-- ram_mat = ram.final  %>%  -->
<!--   column_to_rownames("EnsemblID") -->
<!-- ram.decon = run_dtg(ram_mat , sigsBrain$MB) -->

<!-- write.csv(ram.decon, file= "/home/neuro/Documents/BrainData/Bulk/Ramakar/Formatted/Ramaker-decon.csv") -->
<!-- ``` -->

# Human developmental biology resource 

From recount3, I have also retrieved the dataset from the Human Developmental Biology Resource (HDBR) which contains the largest resource of prenatal samples. 

- Load the data 

```{r}
hdbr = recount3::create_rse_manual(
  project = "ERP016243",
  type = "gene"
)

hdbr_annot = read.csv("../annotations/HDBR-metadata-annot.csv", header=TRUE)
hdbr_supp = read_excel("/home/neuro/Documents/BrainData/Bulk/HDBR/Amended Supplementary Table 1.xlsx", 
                       sheet = 1, skip =6) %>% 
  as.data.frame()

colnames(hdbr_supp) = c("DonorID","Age", "SampleID", "Structure","Hemisphere", "Sex", "PMI")
hdbr_supp %<>% dplyr::select("DonorID","Structure", "PMI")
```

```{r}
hdbr_phenotype = read.csv("/home/neuro/Documents/BrainData/Bulk/HDBR/hdbr-phenotype.csv", 
                          header=TRUE)

```

### Metadata processing 


```{r}
load("../annotations/HDBR.Rda")



md.hdbr = md.hdbr[,1:15]

colnames(md.hdbr) = hdbr_annot$BITColumnName[match(colnames(md.hdbr), hdbr_annot$OriginalMetadataColumnName)]

md.hdbr %<>% 
  as.data.frame() %>% 
  dplyr::select(-c(Block, OntologyIndividual,  KaryotypeOntology, Organism, OrganismOntology, 
                   StructureOntology, HemisphereOntology, OntologyAge)) %>% 
  mutate(AgeInterval = add_feature(.$Age, age_intervals))  %>%
  left_join(strc_acr, by = "Structure") %>% 
  left_join(rgns_ftl, by = "StructureAcronym") %>% 
  mutate(Sex = toupper(Sex)) %>% 
  mutate(Sex = ifelse(str_detect(Sex,"XX"), "F",
                      ifelse(str_detect(Sex, "XY"), "M", "Unknown" )), 
         Diagnosis = c("Control"), 
         Period = c("Prenatal"))

md = colData(hdbr) %>% 
  as.data.frame() %>% 
   dplyr::select(-contains("2"))
colnames(md) = annot$BITColumnName[match(colnames(md), annot$RecountName)]

md.hdbr %<>%
  mutate(Hemisphere = gsub("right", "R", Hemisphere), 
         Hemisphere = gsub("left", "L", Hemisphere), 
         Hemisphere = gsub("frontal", "Frontal", Hemisphere))

hdbr_supp %<>% 
  distinct(DonorID, .keep_all = TRUE) %>% 
  mutate(PMI = ifelse(PMI == "UNKNOWN", NA, PMI))
md.hdbr %<>% 
  left_join(hdbr_supp, by = c("DonorID"))
full.md.hdbr =left_join(md.hdbr, md)
full.md.hdbr %<>%
  mutate(PMI = as.numeric(PMI))


```



```{r}
annot_seq = annot %>% 
  dplyr::filter(Type == "Sequencing metrics")
md.clean = full.md.hdbr 
md.clean = full.md.hdbr %>% select_if(~ !any(is.na(.)))
md.clean = md.clean[vapply(md.clean, function(x) length(unique(x)) > 1, logical(1L))]
```

* Selection of metadata variables for subset 

```{r fig.height=30, fig.width=30}
M = cor(data.matrix(md.clean), use = "complete.obs")
corrplot(M, order='AOE')
```

```{r}
rowMeans(M, na.rm = TRUE) %>% scale() %>% as.data.frame() 

```



### Expression matrix preprocessing 

```{r}
assay(hdbr, "counts") = transform_counts(hdbr, round = TRUE)

hdbr.tpm = recount::getTPM(hdbr)

filter <- rowSums(hdbr.tpm, na.rm = TRUE) > 0.05
hdbr.filter.tpm = hdbr.tpm[filter,]
hdbr.filter.tpm <- thresh(hdbr.filter.tpm)

```


```{r}
hdbr.before = log(hdbr.tpm + 0.01) %>%
    melt() %>%
    dplyr::filter(is.finite(value)) %>% 
    ggplot(aes(x=value, color = as.character( Var2))) +
    geom_density() +
    ggtitle("Before filtering") +
    labs(x = "logTPM", y = "Proportion of Genes") + 
    theme_bw() +
    theme(legend.position = "none")



hdbr.after =log(hdbr.filter.tpm + 0.01) %>%
    melt() %>%
    dplyr::filter(is.finite(value)) %>% 
    ggplot(aes(x=value, color = as.character(Var2))) +
    geom_density() +
    ggtitle("After filtering") +
    labs(x = "logTPM", y = "Proportion of Genes") + 
    theme_bw() +
    theme(legend.position = "none")

grid.arrange(hdbr.before ,hdbr.after, ncol =2)
```

### Deconvolution 

```{r}
hdbr.decon = run_dtg(hdbr.tpm , sigsBrain$MB)
```

- principal component analysis


```{r}


pc = log(hdbr.filter.tpm  + 0.05) %>% 
  t() %>% 
   prcomp(scale = TRUE)

pc$x %>% 
  as.data.frame() %>%
  dplyr::select(PC1, PC2, PC3) %>%
   rownames_to_column("SampleID") %>% 
  left_join(full.md.hdbr, by = "SampleID") %>%
  drop_na(Regions) %>%
 # pivot_wider(names_from = "PC", names_prefix = "PC", values_from = "value") %>% 
  ggplot(
    aes(PC1, PC2, colour = Regions)
  ) +
  geom_point(size=5) + theme_bw() +
   labs(
    x = glue("PC1 ({percent(pc$sdev[[1]]^2 / sum(pc$sdev^2), 0.1)})"),
    y = glue("PC2 ({percent(pc$sdev[[2]]^2 / sum(pc$sdev^2), 0.1)})"),
    colour = "Mitochondrial rate",
    shape = "Structure Acronym"
  ) + theme(legend.position = "top")


 pc$x  %>%
    as.data.frame() %>%
  rownames_to_column("SampleID") %>%
  left_join(full.md.hdbr) %>%
   drop_na(Regions) %>%
    dplyr::select(
        PC1, PC2, PC3, 
        Regions, 
        Age,
        AgeInterval,
        Sex, 
        DonorID,
        IntronicRate, 
        TotalNReads, 
        MappingRate
        
    ) %>% 
   mutate(Regions = as.numeric(as.factor(Regions)),
          Sex = as.numeric(as.factor(Sex)), 
          DonorID = as.numeric(as.factor(DonorID)), 
          AgeInterval= as.numeric(as.factor( AgeInterval)),
          Age= as.numeric(as.factor(Age))) %>%
    cor() %>% 
    corrplot(
        type = "lower", 
        diag = FALSE, 
        addCoef.col = 1, 
        number.cex = 0.75)


```


### Variance partition analysis 


```{r}
form.hdbr <- ~ (1|Age) + (1|StructureAcronym) + (1|Sex) + (1|Hemisphere) + (1|SequencingBatch)+  TotalNReads + MappingRate + (1|Regions) + PMI + DonorID
C.hdbr = canCorPairs(form.hdbr, full.md.hdbr)
plotCorrMatrix(C.hdbr)
```


```{r}
varPar.hdbr <- fitExtractVarPartModel(hdbr.filter.tpm ,form.hdbr ,full.md.hdbr)

```

```{r}
var.genes.hdbr= plotVarPart(varPar.hdbr %>%
    as.data.frame() %>%
    dplyr::select(StructureAcronym, SequencingBatch, Age, everything()))
var.total.hdbr  =plotPercentBars(varPar.hdbr[1:10,] )
grid.arrange(var.genes.hdbr,var.total.hdbr, ncol = 2)

ggsave(var.genes.hdbr, 
       filename = "/home/neuro/Documents/Brain_integrative_transcriptome/Results/Thesis_plots/varPart_hdbr.svg",
       height = 6, 
       width = 8, 
       units = "in")
```


```{r}
varPar.hdbr["ENSG00000081041",] %>%
  melt() %>%
  ggplot(aes(x=variable,y=value)) + geom_bar(stat = "identity",
                                             position = "dodge")
```


### Export data 

```{r}
hdbr.final = hdbr.tpm %>% 
  as.data.frame() %>% 
  rownames_to_column("EnsemblID") %>% 
  mutate_at(.vars = "EnsemblID", .funs = gsub, pattern = "\\.[0-9]*$", replacement = "") 

write.csv(hdbr.final,file = "/home/neuro/Documents/BrainData/Bulk/HDBR/Formatted/HDBR-exp.csv")

write.csv(full.md.hdbr, file = "/home/neuro/Documents/BrainData/Bulk/HDBR/Formatted/HDBR-metadata.csv")

write.csv(varPar.hdbr, file = "/home/neuro/Documents/BrainData/Bulk/HDBR/Formatted/HDBR-varPart.csv")
```

### Deconvolution 

```{r}

hdbr_mat = hdbr.final %>% 
  column_to_rownames("EnsemblID")
hdbr.decon = run_dtg(hdbr_mat , sigsBrain$MB)

write.csv(hdbr.decon, file= "/home/neuro/Documents/BrainData/Bulk/HDBR/Formatted/HDBR-decon.csv")
```


